### ISSUE TITLE
How to configure and run algorithms.

### ISSUE THREAD

**Issue by ajnebro** [jMetal/jMetal#28]
In jMetal there are two ways of configuring and running an algorithm:
a. Creating a runner (or a main) class: `NSGAIIRunner` (or `NSGAII_main`), `SMPSORunner` and so on.
b. Defining a `Settings` class to set the parameter values of the algorithm.

The second alternative was motivated by the need of useing and manipulatin the algorithm settings to define experimental studies. Later, the `Settings` classes allowed us to provide another form of running algorithms by using a `Runner` and a `RunnerC` classes. The first one takes as an argument a problem name, finds its corresponding `Settings` class and executes the algorithm; the second ones takes as an argument a properties file, so there is no need of recompiling the code.

By using the `Runner`/`RunnerC` and settings classes the option a. is not necessary, but until now I decided to keep both approaches, mainly because the runner/main one is the "traditional way".

Some people have given me some ideas, e.g., using JSON files for configuring the solvers. 

The point now is to rethink this issue to provide the jMetal users the most flexible and useful approach.

**Comment by matthieu-vergne** [comment]
Discussing about how to use the measures in #46  lead me to to think back about the parameters management. I completely forgot these discussion here and I am glad that what I wanted to say is already said here. I would just suggest a few modifications. Instead of:
```java
public interface Parameter<T> {
    public void setCurrentValue(T value);
    public T getCurrentValue();
}
```
I would use:
```java
public interface Parameter<T> extends PullMeasure<T>, PushMeasure<T> {
    public void set(T value);
}
```
This way you have a full featured parameter, that you can also be notified about (so even the algorithm can react in real time to parameter changes) and a parameter is also a property of the algorithm, thus a valuable measure to display, so the extensions make sense and you can add the parameters to the `MeasureManager` of the algorithm.

I would also use a `ParameterManager` to allow customization and flexibility regarding the use and configuration of parameters. For instance, some values assigned to a given parameter would be valid only if some other parameters have some specific values too. We could implement inter-parameters constraints in an easier way if we use a manager as intermediary than by dealing with that directly in the concerned parameters.

I do not recommend to revise any `XxxRunner` before the parameters are properly managed. Once you have the parameters and measures, you have generic access to the inputs and outputs, so you can properly use the algorithms as black boxes.

If you want, I can start with these `Parameter` interfaces and generic implementations, like I did with the measures. Now you start to clearly understand the advantages and limitations of the measures, probably you get better the point of the parameters.

**Comment by ajnebro** [comment]
Hmmmm, it is not clear to me why the parameters have to be considered as measures. I'm thinking in some typical ones (population size, maximum number of evaluations, crossover probability, etc).

**Comment by matthieu-vergne** [comment]
You are right. I went a bit too fast in my thinking. I am currently implementing the stuff and finally I chosed to keep parameters and measures completely independent, with the parameters having their own set/get/register/unregister methods. However, because parameters *can* be considered as valuable measures, one could use a `MeasurableParameter` implmentation which implements both parameter and measure methods and make the link between them, so it can be used both as a parameter and a measure. So this choice of using a parameter as a measure or not is let at the implementation level. No dependency exists between the generic interfaces.

**Comment by matthieu-vergne** [comment]
A few updates from my experimental branch: I have removed the register/unregister methods for the parameters, but if you want it you can consider the `MeasurableParameter` implementation. the point was to reduce as much as possible to core features.

Now I am considering the constraints, which is far to be easy. I think I had a nice breakthrough idea with pull/push measures, but I will probably need a bigger one for this one. I already tried to implement stuff like that in a generic way, but I don't think I never succeeded to have a proper generic way to do it. I plan to look at some theories to see how value spaces can be specified (tell me if you have good papers about that), because for now I think about inclusion/exclusion rules but it sill does not help much regarding the numerous ways to include/exclude values from a space (specific values, sets of values, infinite ranges of values) nor in providing ways to exploit space properties to reason efficiently on such inclusion/exclusion (e.g. ordered sets, min-max ranges).

So don't expect to see `Parameter` coming soon in a pull request, because if I remember well you already manage constraints your own way, so there is no point to switch to `Parameter` unless we are able to manage at least as much.

You can anyway look at my experimental branch in my fork if you are curious, but avoid pulling anything from it, because I often rebase stuff, meaning that the history of this branch is completely rewritten (so you should delete your own branch and fetch the new one, no fast forward). But at least there is only few commits, so it should be easy to read.

And if you have ideas, please share. {^_^}

**Comment by matthieu-vergne** [comment]
Small update on the parameters. I think I got a nice idea and I am going ahead with that. You can look at my fork in the `generators` branch for the code, in the `org.uma.jmetal.parameter` package.

# Recap on the `Parameter`

In brief, there is the `Parameter`, which is nothing more than something that you can set/get. It focuses on providing an entry to a given parameter of an algorithm. Then, an algorithm being `Parameterable`, it provides a `ParameterManager`, which gives a generic way to retrieve all the parameters of the algorithm. However, you should still know which values to put, and in particular how to build them.

# The generation problematic

While the choice of the value is anyway a user's decision, knowing how to build it can still be automated. For instance, if the parameter is an integer value, then by saying that you want 5 it should make `set((int) 5)`. If it is an enum with 3 values, by giving you the 3 values you can choose the one you want and it can set it. These kinds of setting are easy, because we deal (i) with primitives like int, long or others, which are single-value types that we can easily transpose in text and vice-versa, or (ii) with completely specified sets of values, like enum, which allows to have an exhaustive listing to select the right one without writting anything. Even managing random values for these cases is not hard, because it is always about choosing a single value in random.

However, if you want to deal with more complex and custom parameters, like a set of mutators which can be each configured with a mutation probability, then you need to know how to create them. In particular, the mutation probability is a real value, so you cannot display an exhaustive list of mutators because it would be unmanageable. Also, you could have designed your mutator by setting the mutation probability within the constructor, or through a setter, and we just cannot guess unless we assume a specific convention, which is rather limiting because both have their own advantages and inconvenients and should be used depending on the context. Moreover, some mutators could have additional parameters because of the highly specific kind of mutation they do. For instance you can imagine a mutator for the TSP for which you can choose whether you swap positions or you reverse completely the chain between these two positions, which makes an additional parameter for this mutator (which is itself a possible value for a parameter of the algorithm). So it can become rather complex.

# The generic generators

The point is that all this information is at the concrete parameter level. It is because a parameter is made to deal with a specific kind of information (e.g. instances of a specific `Mutator` class) that we know what can be a valid value. So we cannot extract the generation semantic from there. Because of that, I have added the `GenerableParameter`, which extends `Parameter` and aims at providing a `ValueGenerator` to help building a value among all the possible (valid) values for the parameter. The point here is that we focus on the generation of a single value, not the description of the space of values, thus it remains focused and we don't spend effort in thinking about broad stuffs.

Now, what does the `ValueGenerator`? If it requires to use non-generic methods to set the values, then it does not help in any way in the generalisation process. The point here is that the `ValueGenerator` only generates the value through `generate()`. Nothing is provided to change this value, it just generates it. For instance, you coud imagine that there is some automatic generator (e.g. random) which does not need to set anything, thus it does not make sense to require anything, like for instance `setListOfPossibleValues()` or `setIndex()`. However, in the specific case where there is the need to configure something... then it means that this generator has parameters. So there is the extension `ParameterableValueGenerator`, which just extends `ValueGenerator` and `Parameterable`.

# Parameter-generator recursivity

So basically, if the generator has no parameters, then it generates the value automatically, otherwise there is parameters to configure. And here is the loop: such a sub-parameter is a `Parameter` instance, and if it is also a `GenerableParameter`, then it has itself a `ValueGenerator`, and so on. Now, where does the loop stops? Basically, when we reach the easy cases like primitives and enum stuff. Small example of a parameter hierarchy (p) with their generators (g) :

`GeneticAlgorithm`
- p1 = mutator -> g1 = produce `Mutator` instance (complex)
	- p1.1 = probability to mutate -> g1.1 = produce 0-1 double (trivial)
	- p1.2 = mutation type -> g1.2 = select within an enum list (trivial)
- p2 = crossover operator -> g2 = produce `CrossOver` instance (complex)
	- p2.1 = probability to reproduce -> g2.1 = produce 0-1 double (trivial)
	- p2.2 = random for the chromosome-cut index -> g2.2 = produce `Random` instance (complex)
		- p2.2.1 = random seed -> g2.2.1 = produce long value (trivial)
- p3 = population size -> g3 = produce positive integer value (trivial)

## Algorithm perspective: provide parameter-specific knowledge

So when you design your algorithm, you also design your parameters, and for each of them the corresponding generator. If the generator should instantiate a complex instance, then you can set its own parameters, which have their own generators, and so on. You stop when you reach trivial enough cases. You should do all of that when you have the information, thus when you are the algorithm designer. I have already implemented stuff, you can look in particular `FiniteDiscreteRangeGenerator` and `FiniteContinuousRangeGenerator`. These classes does no just deal with a single range, they map integers (for discrete) or doubles (for continuous) to the set of values you want to generate. For instance, if you want to generate an integer between 10 and 20, then you use the `FiniteDiscreteRangeGenerator`, which asks you the number of elements (here 11) and a subgenerator to translate a 0-based index (from 0 to 10) into the right integer (from 10 to 20), so basically you get the index and you add 10. If you want to manage something more chaotic, like only the values 3, 4, 9, and 14, then you request to manage 4 values and you provide a subgenerator which maps the 0 to 3 indexes to the right values. Or if you are smarter, you make the list {3, 4, 9, 14} and use the generic factory method `ValueGeneratorFactory.createListBasedGenerator(List<Value> list)` to create a `FiniteDiscreteRangeGenerator` which does exactly that. You can also do it for a list of objects, which allows you to stop the loop earlier if you think that it does not make sense to go deeper.

## User perspective: exploit generic methods

From a user's perspective, what does that mean? It means that for each parameter of the algorithm, you get a `ValueGenerator`. If it is a trivial one, like a `FiniteDiscreteRangeGenerator`, then you can use `getMaxIndex()` to know how many elements are available (whatever they are), set a given index with `setIndex()` and get the corresponding value with `generate()`. You can get for instance only the first and the last one to display the range to the user, or generate a random index between 0 and max and get the coresponding (random) value, and so on. No specific knowledge about the specific parameter is required here because what you face is not the set of values, but the set of integers/doubles which are mapped to them. For another parameter which deals with another set of values you will deal with it the same way (integers or doubles), even if it manages completely different types of objects. Now, in the case where you face a generator for a complex object, then you can (in a generic way) rely on the `ParameterManager` it provides, thus browsing the different parameters and their own generators, and looping on that until you reach these trivial cases, thus never needing any parameter-specific knowledge.

# Summary

As I told you in a previous post in this discussion, what you need to parameter an algorithm is not only which values to provide to each parameter of the algorithm itself (p1, p2, p3), but also which values to provide to the parameters of these values (p1.1, p1.2, p2.1, p2.2), and their own parameters (p2.2.1), and so on. Once you have all of them, then you can build the values in a bottom-up way until you are finally able to set the parameters of the algorithm itself. And here is what we should be able to do in a generic way with these generators. In particular, you can look at the factory for methods to create random generators based on existing generators (it deals with the recursivity too). With that, we should be able to generate a completely randomised setting, even with the most complex parameters.

# Future work

Now, the point is to provide stuff to easily build all of that. For the algorithm designer, there is the factory which already have several methods, but which could surely be improved with others. For the user, we can imagine browsers which, from a given algorithm, get all the "root" parameters and retrieve all the "effective" parameters by reaching all the trivial parameters to set. Then, these parameters can be organised on a GUI, or you can have a command line interaction where the program ask you the values at the deepest levels before to ask you higher levels, until you reach the root parameters. The advantage of these trivial cases is that they can be easily translated in text and vice-versa.

If you think about the configuration files, then you can easily imagine a JSON structure where you map a key (the name of the parameter) to its value represented in a textual way, and for the complex values, simply have an embedded JSON structure with its own key-value pairs, and so on. Thus, changing this setting is only a matter of changing the textual values, letting the structures as-is. For sure, I simplify: depending on how you implement your parameters, such an automatic building of configuration files could make a mess, which is still easy to deal with automatically but humanly unmanageable. But that is another point.

To conclude, I have implemented the concepts (interfaces and classes) but nothing has been tested nor commented. I have just written down what I had in mind. We should be also able to deal with dependencies, but I have not pushed enough the stress to know what I am concretely able to do with what I have made. So it is preliminary stuff, but it seems promising to me.

**Comment by ajnebro** [comment]
Ok, I'll take look to this "short" discussion. After a first and fast read, perhaps it would be a good idea to take a runner class (e.g. NSGAIIRunner) and modify it with the interfaces you have designed to see the result. This could help to properly understand your points.

**Comment by matthieu-vergne** [comment]
Sure, I would like to do something like that. It is still preliminary, so I don't think it is the right time for it. I know it is still incomplete and I am working on that. I just provided the rough idea, to introduce the stuff progressively so it does not come in one block from nowhere. Once I think it has significantly improved, I would make another report.

**Comment by ajnebro** [comment]
Ok. The issue seems quite interesting and deserves time to think about it. For example, the idea of using JSON configuration files is very attractive; some users of jMetal asked of somethink like that to avoid having to recompile everything to run a particular algorithm. We took the approach of using properties files, but this has many limitations.

Another idea to think about is that if an algorithm is configured to solve a, let say, `DoubleProblem`, the possible mutation operators  should be compatible with real encoding, i.e, the mutation parameter should only take a "value" from the set of double mutation operators in the framework. Does this fit with you ideas?

**Comment by matthieu-vergne** [comment]
> Another idea to think about is that if an algorithm is configured to solve a, let say, DoubleProblem, the possible mutation operators should be compatible with real encoding, i.e, the mutation parameter should only take a "value" from the set of double mutation operators in the framework. Does this fit with you ideas?

Kind of, but I don't think it covers, by itself, as much as you seems to say. You need something else behind.

The idea of a generator is that the specific generator should generate valid values for the specific parameter. What is a valid value depends on the parameter-related knowledge which is itself dependent on the algorithm-related knowledge. So basically, this is the algorithm designer which decides what is a valid value. Now, if he wants to use any mutator which fits the double constraint, then he should know how to generate each of them. But if it is possible for other people to provide them (e.g. extending jMetal with libs), then there is no reason for him to know all of them in advance. So in practice, while it should generate valid values, it does not mean it can generate *all* the possible valid values.

If he want to do so, for the example of the mutator based on doubles, then he needs to have a way to retrieve all the implementations which fit, including the ones provided by libs extending jMetal. This kind of thing (retrieving implementations fitting specific constraints) need to be supported by jMetal itself. In such a case, at the framework level you can then provide some kind of class manager which knows what has been provided in the framework and its extensions, and provide this manager globally through some static methods. This way, any generator (or other things) which needs it can get it and retrieve the mutators, adding its own check like implementing a `Mutator<Value>` and ensuring the `Value` is compatible with `Double` (whatever it means), which by itself is not trivial in general in Java, and not always possible (type erasure).

To make it clearer, there is the knowledge of the parameter designer, and generators depend on that, and on the other side the knowledge of the mutator designer, and the mutator depend on that. If the parameter designer want to consider all the compatible mutators, he should have all the required knowledge to do so, which is generally met when one of these conditions holds:
- he is the only mutator designer
- he fixes the convention rules for that parameter and the mutator designers follow + the mutator designers inform the parameter designer about their mutators
- jMetal fixes the convention rules and everyone follow (parameter and mutator designers) + the mutator designers inform jMetal about their mutators + jMetal inform the parameter designer about the mutators

The first condition applies when you work alone, hardly otherwise (unless you are the dedicated person and nobody is allowed to work on that excepted you). The second applies easily when you work with a reduced, known team, but it generally leads to highly specialized code, thus hard to maintain, and if you repeat this kind of logics with several stuffs it add a lot of dependencies. The last one is the most generic, where you break the dependency between a specific parameter and a specific operator by providing generic conventions and managing them at a generic level. Then, it is the generic intermediary which provides all the needed features to work properly at the specific levels (e.g. parameter and operators) and the specific implementations don't have to care about the other specific implementations.

In our case, the whole point for designing a generic, user-friendly framework is to identify:
- the right intermediaries to provide
- the right convention rules to enforce

There could have some flexibility regarding the convention rules, meaning allowing different patterns for the same thing like having an empty constructor or a constructor with specific arguments in a specific order, but they should not overlap in an incompatible way, otherwise we cannot know which convention is actually applied by a given implementation. Once this is done, the specific designers does not care about what other designers do, they only rely on the generic intermediaries by assuming that the conventions concerning them are met.

So while a generator totally *allow* this kind of thing, it does no *support* it. You need something else to bring the stuff implemented by some libs to other libs.
